### PR TITLE
geometry: Add support for boundingbox3d

### PR DESCRIPTION
**Pull request by ptitjano**
## Description

This mostly adds for the relevant geometry classes: 
- boundingBox3D()
- calculateBoundingBox3D()
- boundingBoxIntersects( const QgsBox3D &box3d )

cc @benoitdm-oslandia @lbartoletti

### CODE REVIEW

**Review by benoitdm-oslandia — APPROVED** [review]
Great! Thanks for the tests and the specific fix for combineWith!

**Comment by ptitjano** [comment]
The `QgsBox3D` fixes now have their own PR: https://github.com/qgis/QGIS/pull/53967

**Review by nyalldawson — CHANGES_REQUESTED** [review]
There's a LOT of near-duplicates of the existing 2d code introduced here, and a lot of that logic is non trivial. 

I think we can avoid nearly all of this by always calculating the 3d bounding box **instead** of the 2d box.  I'd suggest we replace all the cached `mutable QgsRectangle mBoundingBox;`  members with their QgsBox3D equivalents, and then adapt the existing bounding box methods to calculate and store the 3d extents instead of 2d.

At worst we'll be incurring a very negligable extra cost of calculating the z ranges for 3d geometries which aren't needed, but that cost is so low that I'd prefer it over the extra code paths. And if it proves to be a problem, then I'd only suggest we do optimised separate 2d/3d paths for the simple cases like line strings (not curved types).

**Comment by ptitjano** [comment]
> There's a LOT of near-duplicates of the existing 2d code introduced here, and a lot of that logic is non trivial.
> 
> I think we can avoid nearly all of this by always calculating the 3d bounding box **instead** of the 2d box. I'd suggest we replace all the cached `mutable QgsRectangle mBoundingBox;` members with their QgsBox3D equivalents, and then adapt the existing bounding box methods to calculate and store the 3d extents instead of 2d.
> 
> At worst we'll be incurring a very negligable extra cost of calculating the z ranges for 3d geometries which aren't needed, but that cost is so low that I'd prefer it over the extra code paths. And if it proves to be a problem, then I'd only suggest we do optimised separate 2d/3d paths for the simple cases like line strings (not curved types).

I have removed all the duplicated 2d code, except for `QgsPoint` : it does not seem appropriate because of the corner checks which are really different between the 2d and 3d cases.

I don't understand the `mutable QgsRectangle mBoundingBox;` comment. Which ones are you talking about?

**Comment by ptitjano** [comment]
@nyalldawson gentle ping

File: src/core/geometry/qgslinestring.cpp [hunk 1]
```diff
@@ -502,6 +502,93 @@ bool QgsLineString::boundingBoxIntersects( const QgsRectangle &rectangle ) const
   return QgsCurve::boundingBoxIntersects( rectangle );
 }
 
+bool QgsLineString::boundingBoxIntersects( const QgsBox3D &box3d ) const
+{
+  if ( mX.empty() )
+    return false;
+
+  if ( mZ.empty() )
+    return boundingBoxIntersects( QgsRectangle( box3d.xMinimum(), box3d.yMinimum(), box3d.xMaximum(), box3d.yMaximum() ) );
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp:511]:
```suggestion
    return boundingBoxIntersects( box3d.toRectangle() );
```

File: src/core/geometry/qgslinestring.cpp [hunk 2]
```diff
@@ -502,6 +502,93 @@ bool QgsLineString::boundingBoxIntersects( const QgsRectangle &rectangle ) const
...
+
+  if ( !mBoundingBox.isNull() )
+  {
+    return mBoundingBox.intersects( box3d );
+  }
+  const int nb = mX.size();
+
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp:518]: Given this is near duplicate code to `QgsLineString::boundingBoxIntersects( const QgsRectangle &rectangle )`, can you add some comments to both that "if one of these functions is changed then remember to also update the other accordingly".

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Just a couple of minor comments, otherwise all looks good thanks!

File: src/core/geometry/qgslinestring.cpp (hunk 1 above)
Comment (benoitdm-oslandia) [reply src/core/geometry/qgslinestring.cpp]: ok

File: src/core/geometry/qgslinestring.cpp (hunk 2 above)
Comment (benoitdm-oslandia) [reply src/core/geometry/qgslinestring.cpp:522]: ok

**Review by benoitdm-oslandia — APPROVED** [review]

**Pull request merged**